pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,335 @@
+         IsProcessRunning(babysitter_pid_)) {
+    timer_->SleepUs(30 * 1000 /* 0.03s */);
+  }
+
+  if (!IsProcessRunning(master_pid_)) {
+    handler_->Message(
+        kInfo, "Master process (%d) exited; controller shutting down.",
+        master_pid_);
+  } else {
+    handler_->Message(
+        kInfo, "Babysitter process (%d) exited; controller shutting down.",
+        babysitter_pid_);
+  }
+
+  // TODO(jefftk): Once the controller is doing real work and there's real
+  // cleanup work to do, we can initiate that from here.
+
+  exit(EXIT_SUCCESS);
+
+  // Because the controller is exiting via exit(), the babysitter process will
+  // see that and quit itself instead of restarting it.
+}
+
+pid_t ControllerManager::ReadPidFromFile(
+    const GoogleString& filename) {
+  if (file_system_->Exists(
+          filename.c_str(), handler_).is_true()) {
+    GoogleString pid_buffer;
+    if (file_system_->ReadFile(
+            filename.c_str(), &pid_buffer, handler_)) {
+      GoogleString pid_buffer_trimmed;
+      TrimWhitespace(pid_buffer, &pid_buffer_trimmed);
+
+      int64 pid64;
+      if (StringToInt64(pid_buffer_trimmed, &pid64)) {
+        pid_t pid = static_cast<pid_t>(pid64);
+        return pid;
+      }
+    }
+  }
+  return -1;  // Failed to read pid.
+}
+
+pid_t ControllerManager::MasterPid() {
+  GoogleString pid_filename = factory_->MasterPidFilename();
+
+  // If the file doesn't exist, or contains the PID of something that isn't
+  // running, block until we have a usable PID.  This can happen if the host
+  // webserver is taking a little while to finish daemonizing.
+  for (int i = 1 /* 1 so we don't warn right away */; true; i++) {
+    pid_t pid = ReadPidFromFile(pid_filename);
+    if (pid != -1 && IsProcessRunning(pid)) {
+      return pid;
+    }
+
+    if (i % 100 == 0) {
+      handler_->Message(
+          kWarning,
+          "Controller stuck waiting for master process PID to appear in %s",
+          pid_filename.c_str());
+    }
+
+    // Sleep instead of spinning.
+    timer_->SleepUs(10 * 1000 /* 0.01s */);
+  }
+}
+
+GoogleString ControllerManager::ControllerPidFilename() {
+  SystemRewriteOptions* options = reinterpret_cast<SystemRewriteOptions*>(
+      factory_->default_options());  // Cast is safe, since we're in system.
+  CHECK(options != NULL);
+
+  GoogleString controller_pid_filename(options->controller_pid_filename());
+  if (controller_pid_filename.empty()) {
+    handler_->Message(
+        kError, "Controller PID Filename should not be set to empty");
+    return "";
+  } else if (controller_pid_filename[0] == '/') {
+    return controller_pid_filename;
+  } else {
+    // Relative paths are interpreted relative to the directory in which the
+    // webserver would normally store the pid file for its master process.
+
+    GoogleString master_pid_filename = factory_->MasterPidFilename();
+    size_t final_slash = master_pid_filename.find_last_of('/');
```

Comment:
`dirname` is evil.  It can modify it's input, segfault on static strings, and isn't thread safe.